File: src/main/java/net/kimleo/rec/v2/accessor/Accessor.java
package net.kimleo.rec.v2.accessor;

import net.kimleo.rec.Pair;
import net.kimleo.rec.v2.accessor.lexer.Lexer;
import net.kimleo.rec.concept.Accessible;

import java.util.Map;

import static net.kimleo.rec.exception.Assert.assertTrue;

public class Accessor<T> {

    private final Map<String, Integer> fieldMap;
    private final Integer leastCapacity;

    public Accessor(String[] fields) {
        Pair<Map<String, Integer>, Integer> fieldMapPair = Lexer.buildFieldMapPair(fields);
        fieldMap = fieldMapPair.getFirst();
        leastCapacity = fieldMapPair.getSecond();
    }

    public Map<String, Integer> getFieldMap() {
        return fieldMap;
    }

    public Integer getLeastCapacity() {
        return leastCapacity;
    }

    public RecordWrapper<T> create(Accessible<T> record) {
        assertTrue (record.size() >= leastCapacity);

        return new RecordWrapper<>(fieldMap, record);
    }

    public RecordWrapper<T> of(Accessible<T> record) {
        return create(record);
    }
}

File: rec-accessor/src/main/kotlin/net/kimleo/rec/accessor/Accessor.kt
package net.kimleo.rec.accessor

import net.kimleo.rec.accessor.lexer.Lexer
import net.kimleo.rec.concept.Indexible

class Accessor<T>(fields: Array<out String>) {
    val fieldMapPair = Lexer.buildFieldMapPair(*fields)
    val fieldMap = fieldMapPair.first
    val leastCapacity = fieldMapPair.second

    fun create(record: Indexible<T>): RecordWrapper<T> {
        assert(record.size >= leastCapacity)

        return RecordWrapper(fieldMap, record)
    }

    fun of(record: Indexible<T>) = create(record)
}